flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectOfTypeStrategy.java

```diff
@@ -0,0 +1,129 @@
+ * </ul>
+ *
+ * <p>Field type normalization includes:
+ *
+ * <ul>
+ *   <li>Converting CHARACTER_STRING types (CHAR, VARCHAR) to nullable VARCHAR (STRING)
+ *   <li>Making all field types nullable for flexibility in structured types
+ * </ul>
+ *
+ * <p>The strategy returns {@code Optional.empty()} if the class name argument is not available as a
+ * literal value (e.g., during type inference testing scenarios).
+ *
+ * <p><b>Examples:</b>
+ *
+ * <ul>
+ *   <li>{@code OBJECT_OF('com.example.User', 'name', 'Alice', 'age', 30)} → {@code
+ *       STRUCTURED<com.example.User>(name STRING, age INT)}
+ *   <li>{@code OBJECT_OF('com.example.Point', 'x', 1.5, 'y', 2.0)} → {@code
+ *       STRUCTURED<com.example.Point>(x DOUBLE, y DOUBLE)}
+ * </ul>
+ *
+ * <p><b>Implementation Notes:</b>
+ *
+ * <ul>
+ *   <li>Field names must be available as string literals during type inference
+ *   <li>The class name is used for type identification but the runtime representation is RowData
+ *   <li>Uses {@link IntStream} for efficient processing of key-value pairs
+ * </ul>
+ *
+ * @see org.apache.flink.table.functions.BuiltInFunctionDefinitions#OBJECT_OF
+ * @see ObjectOfInputTypeStrategy
+ */
+public class ObjectOfTypeStrategy implements TypeStrategy {
+
+    private static DataType toStructuredType(
+            final String className, final CallContext callContext) {
+        final List<DataType> argumentDataTypes1 = callContext.getArgumentDataTypes();
+        final DataTypes.Field[] fields =
+                IntStream.range(1, argumentDataTypes1.size())
+                        .filter(i -> i % 2 == 1)
+                        .mapToObj(keyIdx -> toFieldDataType(callContext, keyIdx))
+                        .toArray(DataTypes.Field[]::new);
+
+        return DataTypes.STRUCTURED(className, fields);
+    }
+
+    private static Field toFieldDataType(final CallContext callContext, final int keyIdx) {
+        final List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
+
+        final String fieldName =
+                callContext
+                        .getArgumentValue(keyIdx, String.class)
+                        .orElseThrow(IllegalStateException::new);
+
+        final DataType fieldValueType = argumentDataTypes.get(keyIdx + 1);
+        final DataType normalizedFieldValueType = normalizeFieldType(fieldValueType);
+        return DataTypes.FIELD(fieldName, normalizedFieldValueType);
+    }
+
+    /**
+     * Normalizes field types for structured types: - Converts CHARACTER_STRING types (CHAR,
+     * VARCHAR) to VARCHAR (STRING) - Makes all types nullable for flexibility in structured types.
+     */
+    private static DataType normalizeFieldType(DataType fieldType) {
+        final LogicalType logicalType = fieldType.getLogicalType();
+
+        if (logicalType.is(LogicalTypeFamily.CHARACTER_STRING)) {
+            return DataTypes.STRING().nullable();
+        }
```

Comment:
@twalthr I was not 100% that this is the correct approach! I noticed in the output of the tests that strings were used as `CHAR(n)`, so I had to convert them to `STRING()`. I couldn't find any util class that does this conversion, so I added the logic here.